### PR TITLE
[core] document unique field validation per content type and locale

### PR DESCRIPTION
**Pull request by jhoward1994**
### What does it do?

Validates document level unique fields according to the following rules:
- check uniqueness only across documents in the same locale, publication state and content-type and not across locales/publication states within the same document
- Adds e2e and api tests for all types of document unique fields

### Why is it needed?

Fixes unique field validation rules for v5

### How to test it?

- Using the CM test document unique field behaviour adheres to the above rules
- Run the modified/added tests in this PR, e.g:
```
yarn test:e2e --setup --domains=content-manager
LICENSE_INFO yarn test:api api-tests/core/strapi/document-service
LICENSE_INFO yarn test:api api-tests/plugins/i18n
yarn test:unit packages/core/core/src/services/
yarn test:front packages/core/admin/admin/src/content-manager/
```

### Related issue(s)/PR(s)

CONTENT-2117

### CODE REVIEW

File: packages/core/core/src/services/entity-validator/validators.ts [hunk 1]
```diff
@@ -163,13 +170,15 @@ const addUniqueValidator = <T extends strapiUtils.yup.AnySchema>(
       return true;
     }
 
-    const whereParams = entity
-      ? { $and: [{ [updatedAttribute.name]: value }, { $not: { id: entity.id } }] }
+    /**
+     * At this point we know that we are creating a new entry or that the unique field value has changed
+     * We check if there is an entry of this content type in the same locale with the same unique field value
+     */
```
Comment (Marc-Roig) [inline packages/core/core/src/services/entity-validator/validators.ts:176]: Shouldn't updates of an entity get here too?

File: packages/core/core/src/services/entity-validator/validators.ts [hunk 2]
```diff
@@ -138,7 +139,13 @@ const addStringRegexValidator = (
...
 const addUniqueValidator = <T extends strapiUtils.yup.AnySchema>(
   validator: T,
-  { attr, model, updatedAttribute, entity }: ValidatorMetas<Attribute.Any & Attribute.UniqueOption>
+  {
+    attr,
+    model,
+    updatedAttribute,
+    entity,
+    locale,
+  }: ValidatorMetas<Attribute.Any & Attribute.UniqueOption>
```
Comment (Marc-Roig) [inline packages/core/core/src/services/entity-validator/validators.ts:148]: I am in the process of manually testing this, but I believe we forgot about the publication state of the document.

Fields should be unique across the same locale, but also across the same publication state.

**Review by Marc-Roig — COMMENTED** [review]
Good job :) 

I will be manually QAing this.

In the meantime, if possible I would add api tests to validate the uniqueness on some edge cases.

I was thinking of adding them in:
`api-tests/core/strapi/document-service` // Test publication state uniqueness, and overall unique values work with doc service
`api-tests/plugins/i18n` // Test locale uniqueness

I would give extra time and attention to tests, as we have seen some issues on uniqueness before.

Also, we should also make fields unique accross the same publication state, draft and published versions of a document should be able to have the same values on unique fields.

**Comment by Marc-Roig** [comment]
I believe uniqueness could have it's own blitz session with the team, wdyt? when v5/draft-and-publish branch is merged (hopefully during the week) we could mention this to the team if you see it appropiate!

**Review by Marc-Roig — COMMENTED** [review]
Just QAd this branch, everything works as expected :) 
I will wait for the CM fix or the api tests (or any update you see fit)!

File: api-tests/plugins/i18n/content-manager/crud.test.api.js [hunk 3]
```diff
@@ -111,20 +112,27 @@ describe('i18n - Content API', () => {
           method: 'POST',
           url: `/content-manager/collection-types/api::category.category`,
           qs: {
-            plugins: { i18n: { locale, relatedEntityId: data.categories[0].id } },
```
Comment (Marc-Roig) [inline api-tests/plugins/i18n/content-manager/crud.test.api.js:114]: do you know in what way we were using the `relatedEntityId` . I actually forgot why we had that 🤔

File: api-tests/core/strapi/document-service/resources/schemas/author.js [hunk 4]
```diff
@@ -10,15 +10,16 @@ module.exports = {
...
+      localized: true,
     },
   },
   attributes: {
     name: {
       type: 'string',
+      unique: true,
       pluginOptions: {
         i18n: {
-          localized: false,
```
Comment (Marc-Roig) [inline api-tests/core/strapi/document-service/resources/schemas/author.js:21]: other tests depend on this to be non localized , would it be possible to use another schema here? (category) 
Maybe I should add a comment actually mentioning this 🤔

File: api-tests/core/strapi/document-service/unique.test.api.ts [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        data: { name: newValueToTest },
+      });
+
+      expect(author2).toMatchObject({
+        name: newValueToTest,
+        publishedAt: null,
+      });
+
+      const publishRes2 = strapi.documents(AUTHOR_UID).publish(author2.documentId);
+      await expect(publishRes2).resolves.not.toThrowError();
```
Comment (Marc-Roig) [inline api-tests/core/strapi/document-service/unique.test.api.ts:56]: I would also validate the uniqueness works on publish and update of a document, 
so if you try to publish a document and the unique field is duplicated in the published version, it should throw an error.

File: e2e/tests/content-manager/editview.spec.js [hunk 6]
```diff
@@ -24,7 +21,7 @@ test.describe('Edit View', () => {
     /**
      * Now we're in the edit view.
      */
-    await page.waitForURL('**/content-manager/collection-types/api::testing.testing/create');
+    await page.waitForURL('**/content-manager/collectionType/api::testing.testing/create?**');
```
Comment (Marc-Roig) [inline e2e/tests/content-manager/editview.spec.js:24]: just curious, why the `?**` ?

File: packages/core/core/src/services/document-service/index.ts [hunk 7]
```diff
@@ -131,7 +131,10 @@ const createDocumentService = ({
 
     const model = strapi.getModel(uid) as Shared.ContentTypes[Common.UID.ContentType];
 
-    const validData = await entityValidator.validateEntityCreation(model, data, { isDraft: true });
+    const validData = await entityValidator.validateEntityCreation(model, data, {
+      isDraft: true,
```
Comment (Marc-Roig) [inline packages/core/core/src/services/document-service/index.ts:135]: I just realized, what about publishing a document? Shouldn't it be validating a publsihed version entity creation?

File: packages/core/core/src/services/entity-validator/validators.ts [hunk 8]
```diff
@@ -163,13 +166,20 @@ const addUniqueValidator = <T extends strapiUtils.yup.AnySchema>(
...
+          locale: options.locale,
+        }
       : { [updatedAttribute.name]: value };
 
-    const record = await strapi.query(model.uid).findOne({
-      select: ['id'],
-      where: whereParams,
+    const record = await strapi.documents.findFirst(model.uid, {
+      filters,
+      status: options.isDraft ? 'draft' : 'published',
```
Comment (Marc-Roig) [inline packages/core/core/src/services/entity-validator/validators.ts:182]: Preferably, the locale should be placed as a top level attribute :
```suggestion
    const record = await strapi.documents.findFirst(model.uid, {
      filters: { [updatedAttribute.name]: value },
      locale: options.locale
      status: options.isDraft ? 'draft' : 'published',
```

File: api-tests/plugins/i18n/content-manager/crud.test.api.js (hunk 3 above)
Comment (jhoward1994) [reply api-tests/plugins/i18n/content-manager/crud.test.api.js:114]: see my comment here [`1c6897c` (#19153)](https://github.com/strapi/strapi/pull/19153/commits/1c6897cdf8a8b1ce4ff479da57d3265db14c3004)

wdyt?

Comment (Marc-Roig) [reply api-tests/plugins/i18n/content-manager/crud.test.api.js:114]: makes sesnse :)

File: e2e/tests/content-manager/editview.spec.js [hunk 9]
```diff
@@ -26,7 +23,7 @@ test.describe('Edit View', () => {
     /**
      * Now we're in the edit view.
      */
-    await page.waitForURL('**/content-manager/collection-types/api::testing.testing/create');
+    await page.waitForURL('**/content-manager/collection-types/api::testing.testing/create?');
```
Comment (joshuaellis) [inline e2e/tests/content-manager/editview.spec.js:26]: how come you needed to change this? :)

File: e2e/tests/content-manager/uniqueness.spec.js [hunk 10]
```diff
@@ -0,0 +1,110 @@
+import { test, expect } from '@playwright/test';
+import { login } from '../../utils/login';
+import { resetDatabaseAndImportDataFromPath } from '../../scripts/dts-import';
+
+test.describe('Uniqueness', () => {
+  test.beforeEach(async ({ page }) => {
+    // Reset the DB and also specify that we are wiping all entries of the testing content type each time
```
Comment (joshuaellis) [inline e2e/tests/content-manager/uniqueness.spec.js:7]: Are they not wiped already?

File: packages/core/core/src/services/entity-validator/index.ts [hunk 11]
```diff
@@ -315,7 +316,11 @@ const createValidateEntity = (createOrUpdate: CreateOrUpdate) => {
 
     const validator = createModelValidator(createOrUpdate)(
       { model, data, entity },
-      { isDraft: options?.isDraft ?? false }
+      {
+        isDraft: options?.isDraft ?? false,
+        // TODO should we pass the default locale here?
+        locale: options?.locale ?? 'en',
```
Comment (Marc-Roig) [inline packages/core/core/src/services/entity-validator/index.ts:322]: locale should be null by default if not provided (locale is null for content types with disabled i18n). Whoever is using this the entity validator should provide the appropiate locale, and if not I would say it should be empty by default.

File: packages/core/core/src/services/entity-validator/validators.ts [hunk 12]
```diff
@@ -138,18 +140,24 @@ const addStringRegexValidator = (
...
 ): T => {
   if (attr.type !== 'uid' && !attr.unique) {
     return validator;
   }
 
   return validator.test('unique', 'This attribute must be unique', async (value) => {
+    const isPublish = options.isDraft === false;
+
+    // When publishing, the value will not have changed so we take the value from the entity
+    const valueToCheck = isPublish ? entity?.[updatedAttribute.name] : value;
```
Comment (Marc-Roig) [inline packages/core/core/src/services/entity-validator/validators.ts:154]: whats the difference of taking it from the entity or from the value when its published? Is the "value" empty when publishing?

File: e2e/tests/content-manager/uniqueness.spec.js (hunk 10 above)
Comment (jhoward1994) [reply e2e/tests/content-manager/uniqueness.spec.js:7]: No they're not, I found when moving between browsers the previous documents were still there and unique constraints would fail. I think this hasn't been a problem in other specs as we haven't been testing constraints or we've been deleting the entry within the user flow(?). 

Do you think this is the best way to handle this? Or should we find a way for `resetDatabaseAndImportDataFromPath` to clear all entries by default?

File: e2e/tests/content-manager/editview.spec.js (hunk 9 above)
Comment (jhoward1994) [reply e2e/tests/content-manager/editview.spec.js]: I've actually found the test gets stuck here unless we add `?**` to match all query params in the URL

File: packages/core/core/src/services/entity-validator/validators.ts (hunk 12 above)
Comment (jhoward1994) [reply packages/core/core/src/services/entity-validator/validators.ts:154]: Yes the value is empty when publishing but the entity will be populated

**Comment by jhoward1994** [comment]
@Marc-Roig thanks for the review. wdyt of the latest changes?

There was a bug with publishing non default locales which I've fixed and I think we should talk about the behaviour for UID fields as our current definition violates a DB constraint

File: packages/core/core/src/services/document-service/index.ts (hunk 7 above)
Comment (jhoward1994) [reply packages/core/core/src/services/document-service/index.ts]: Could you elaborate?

Comment (Marc-Roig) [reply packages/core/core/src/services/document-service/index.ts]: when publishing a draft version, we need the validations to apply to the published versions , for example:

Given a content type with a unique field named code

Doc1-Draft  { code: 1 }

Doc2-Draft { code: 2 }
Doc2-Published { code :  1}

If you were to publish Doc1-Draft , shouldn't you get an error saying code is a unique attribute and that value already exist in Doc2-Published?

Side question, we want that behaviour right?

**Comment by Marc-Roig** [comment]
@jhoward1994 Changes look good to me :) I am still concerned about unique values accross publication states, and how that should work. I left a comment about that

File: packages/core/core/src/services/document-service/index.ts (hunk 7 above)
Comment (jhoward1994) [reply packages/core/core/src/services/document-service/index.ts]: Yep thats the right behaviour and I think it's what we have atm, see video. Is there something I've missed?


https://github.com/strapi/strapi/assets/48524071/ddf2a04b-dc68-43f7-8bcf-02e559cd64f9

_I should have just spoken over this video sorry, hope it's clear!_

Comment (Marc-Roig) [reply packages/core/core/src/services/document-service/index.ts]: oh :O , then I wonder how i tworks. Why do we have an "isDraft" attribute that that is always true? that parameter is never sent as false. So I wonder at which point we do this validation for published versions.

Comment (jhoward1994) [reply packages/core/core/src/services/document-service/index.ts]: Publishing comes through the document-engine.clone here - https://github.com/strapi/strapi/blob/81fd684402ab62b66eba947c1af8a71a333b3621/packages/core/core/src/services/document-service/document-engine.ts#L227

where isDraft is set. We only default to true in the create & update e.g. https://github.com/strapi/strapi/blob/81fd684402ab62b66eba947c1af8a71a333b3621/packages/core/core/src/services/document-service/document-engine.ts#L177 which I think is correct behaviour?

File: packages/plugins/i18n/server/src/controllers/validate-locale-creation.ts [hunk 13]
```diff
@@ -26,6 +26,9 @@ const validateLocaleCreation: Common.MiddlewareHandler = async (ctx, next) => {
 
   // Prevent empty string locale
   const locale = get('locale', query) || get('locale', body) || undefined;
+
+  // TODO:
+  // const relatedEntityId = get('plugins.i18n.relatedEntityId', query);
```
Comment (Marc-Roig) [inline packages/plugins/i18n/server/src/controllers/validate-locale-creation.ts:31]: should we just remove this?

**Review by Marc-Roig — APPROVED** [review]
great job jamie 🚀
